flink-ml-lib/src/main/java/org/apache/flink/ml/classification/knn/KnnModel.java

```diff
@@ -0,0 +1,217 @@
+        private KnnModelData knnModelData;
+        private final int k;
+        private final String broadcastKey;
+
+        public PredictLabelFunction(String broadcastKey, int k, String featureCol) {
+            this.k = k;
+            this.broadcastKey = broadcastKey;
+            this.featureCol = featureCol;
+        }
+
+        @Override
+        public Row map(Row row) {
+            if (knnModelData == null) {
+                knnModelData =
+                        (KnnModelData)
+                                getRuntimeContext().getBroadcastVariable(broadcastKey).get(0);
+            }
+            DenseVector feature = (DenseVector) row.getField(featureCol);
+            double prediction = predictLabel(feature);
+            return Row.join(row, Row.of(prediction));
+        }
+
+        private double predictLabel(DenseVector feature) {
+            PriorityQueue<Tuple2<Double, Double>> priorityQueue =
+                    new PriorityQueue<>(Comparator.comparingDouble(o -> -o.f0));
+            double featureNorm = 0.0;
+            for (int i = 0; i < feature.size(); ++i) {
+                featureNorm += feature.values[i] * feature.values[i];
+            }
+            Tuple2<DenseVector, Double> featureAndNorm = Tuple2.of(feature, featureNorm);
+            double[] labelValues = knnModelData.labels.values;
+            double[] normValues = knnModelData.featureNorms.values;
+            DenseVector distanceVector = new DenseVector(labelValues.length);
+            BLAS.gemv(
+                    -2.0,
+                    knnModelData.packedFeatures,
+                    true,
+                    featureAndNorm.f0,
+                    0.0,
+                    distanceVector);
+            for (int i = 0; i < distanceVector.values.length; i++) {
+                distanceVector.values[i] =
+                        Math.sqrt(
+                                Math.abs(
+                                        distanceVector.values[i]
+                                                + featureAndNorm.f1
+                                                + normValues[i]));
+            }
+            List<Tuple2<Double, Double>> distanceAndLabels = new ArrayList<>(labelValues.length);
+            for (int i = 0; i < labelValues.length; i++) {
+                distanceAndLabels.add(Tuple2.of(distanceVector.values[i], labelValues[i]));
+            }
+            for (Tuple2<Double, Double> distanceAndLabel : distanceAndLabels) {
+                if (priorityQueue.size() < k) {
+                    priorityQueue.add(distanceAndLabel);
+                } else {
+                    Tuple2<Double, Double> head = priorityQueue.peek();
+                    if (head.f0 > distanceAndLabel.f0) {
+                        priorityQueue.poll();
+                        priorityQueue.add(distanceAndLabel);
+                    }
+                }
+            }
+            List<Double> labels = new ArrayList<>();
```

Comment:
nits: Can you remove the `labels` here and do the weight aggregation directly? It seems that some code is redundant.

Reply:
OK